torch/csrc/autograd/engine.cpp

```diff
@@ -289,51 +326,29 @@ auto Engine::thread_main(
     // The outer graph_task represents the overall graph_task we need to execute
     // for reentrant execution.
     std::shared_ptr<GraphTask> local_graph_task;
-    if (!(local_graph_task = task.base_.lock())) {
+    if (!(local_graph_task = task->base_.lock())) {
       // Reentrant thread's graph task should not expire since we hold a
       // reference to it in this method.
       TORCH_INTERNAL_ASSERT(!reentrant_thread);
-      LOG(INFO) << "GraphTask for function " << task.fn_->name()
+      LOG(INFO) << "GraphTask for function " << task->fn_->name()
                 << " is no longer valid, skipping execution";
       continue;
     }
 
-    if (task.fn_ && !local_graph_task->has_error_.load()) {
+    if (task->fn_ && !local_graph_task->has_error_.load()) {
       GradMode::set_enabled(local_graph_task->grad_mode_);
       try {
-        evaluate_function(local_graph_task, task.fn_.get(), task.inputs_);
+        evaluate_function(local_graph_task, task->fn_.get(), task->inputs_);
       } catch (std::exception& e) {
-        thread_on_exception(local_graph_task, task.fn_, e);
+        thread_on_exception(local_graph_task, task->fn_, e);
       }
     }
-    // Notify downstream about the completion of tasks depending
-    // on both where the task was executed, and who owned the overall
-    // graph (in case of reentrant execution.)  See Note [Reentrant backwards].
-    auto base_owner = local_graph_task->owner_;
-    // Task from a non-worker thread. Easy case.
-    if (base_owner == NO_DEVICE) {
-      if (--local_graph_task->outstanding_tasks_ == 0) {
-        // Lock mutex to notify the GraphTask waiting on not_done_
-        std::lock_guard<std::mutex> lock(local_graph_task->mutex_);
-        local_graph_task->not_done_.notify_all();
-      }
-    } else {
-      // If it's a task initiated from this thread, decrease the counter, but
-      // don't do anything - loop condition will do all checks for us next.
-      if (base_owner == worker_device) {
-        --local_graph_task->outstanding_tasks_;
-        // Otherwise send a dummy function task to the owning thread just to
-        // ensure that it's not sleeping. If it has work, it might see that
-        // graph_task->outstanding_tasks_ == 0 before it gets to the task, but
-        // it's a no-op anyway.
-      } else if (base_owner != worker_device) {
-        if (--local_graph_task->outstanding_tasks_ == 0) {
-          // Synchronize outstanding_tasks_ with queue mutex
-          std::atomic_thread_fence(std::memory_order_release);
-          ready_queue_by_index(base_owner)
-              .push(NodeTask(local_graph_task, nullptr, InputBuffer(0)));
-        }
-      }
+
+    // Notify downstream about the completion of tasks
+    if (--local_graph_task->outstanding_tasks_ == 0) {
+      // Lock mutex to notify the GraphTask waiting on not_done_
+      std::lock_guard<std::mutex> lock(local_graph_task->mutex_);
+      local_graph_task->not_done_.notify_all();
```

Comment:
Why do you remove all the logic associated to the case where the owner of the task is a worker thread?

Reply:
I think this should be clearly explained somewhere. It's hard to follow that this is correct because of some changes done at some other place in the code.